src/IO/ReadHelpers.h

```diff
@@ -833,6 +837,87 @@ inline bool tryReadUUIDText(UUID & uuid, ReadBuffer & buf)
     return readUUIDTextImpl<bool>(uuid, buf);
 }
 
+template <typename ReturnType = void>
+inline ReturnType readIPv4TextImpl(IPv4 & ip, ReadBuffer & buf)
+{
+    const char * end = parseIPv4(buf.position(), buf.position() + buf.available(), reinterpret_cast<unsigned char *>(&ip.toUnderType()));
```

Comment:
It's not completely correct. If string with ip is on the border of buffer it won't be parsed correctly.

Reply:
what do you mean "ip is on the border of buffer"?

Reply:
buffer always contains complete literal at least - how else would you treat let's say "127.0.0.1"? maybe it's "127.0.0.100"...

Reply:
> buffer always contains complete literal at least

But how is it guaranteed? This function is used inside serialization and there are no literals. You can have arbitrary text sequence and exactly this function have to extract value from raw data (like `readIntText`).

Reply:
apparently it is guaranteed by an expression parser which calls this function - `readIntTextImpl` does exactly the same - reads text until not eof and convertible to number

Reply:
It's guaranteed by expression parser, but it's not guaranteed in other places where this function may be called.

> readIntTextImpl does exactly the same

No. It reads until `eof` or until parsing will knowly fail (non-digit char or overflow). Wherein `eof` calls `ReadBuffer::next()`,  which request more data if we reached the end of working buffer.

In general this test shall pass:
```sql
DROP TABLE IF EXISTS test_file_ip;

CREATE TABLE test_file_ip
(
    `id` UInt32,
    `ip` IPv4,
    `s` String
)
ENGINE = File(JSONEachRow, 'test_file_ip.json');

SET engine_file_truncate_on_insert = 1;
INSERT INTO test_file_ip VALUES(1, toIPv4('190.1.5.64'), 'foo');

SELECT *
FROM file('test_file_ip.json', 'JSONEachRow', 'id UInt32, ip IPv4, s String')
SETTINGS input_format_parallel_parsing = 0;

SELECT *
FROM file('test_file_ip.json', 'JSONEachRow', 'id UInt32, ip IPv4, s String')
SETTINGS input_format_parallel_parsing = 0, max_read_buffer_size = 4;

```

and return

```
1       190.1.5.64      foo
1       190.1.5.64      foo
```

Now it returns:
```
1       190.1.5.64      foo
Received exception from server (version 22.12.1):
Code: 671. DB::Exception: Received from localhost:9000. DB::ParsingException. DB::ParsingException: Cannot parse IPv4 19: (while reading the value of key ip): While executing JSONEachRowRowInputFormat: While executing File: (in file/uri /home/ubuntu/data22/user_files/test_file_ip.json): (at row 1)
. (NAMED_COLLECTION_IS_IMMUTABLE)
(query: SELECT *
FROM file('test_file_ip.json', 'JSONEachRow', 'id UInt32, ip IPv4, s String')
SETTINGS input_format_parallel_parsing = 0, max_read_buffer_size = 4;)
```

Reply:
oh, I missed this eof() functionality... will fix it, thanks